File: results/RQ1/05_baseline_qualification/units/p11_custom_press_releases/java/attempts/0001_seed_2810241787/p11_custom_press_releases_java/p11_custom_press_releases_java/src/main/java/PressRelease.java
/**
 * Represents a press release with associated metadata, content, and lifecycle management.
 * The press release follows a state machine: DRAFT -> REVIEW -> PUBLISHED.
 * It supports custom fields, tagging, version history, and rendering.
 */
public final class PressRelease {
    private final String id;
    private final String headline;
    private final String body;
    private final String contact;
    private final Template template;
    private final java.util.Map<String, String> customFields;
    private ReleaseState state;
    private final java.util.Set<String> tags;
    private final java.util.List<Version> versionHistory;

    /**
     * Constructs a new PressRelease with the specified parameters.
     *
     * @param id the unique identifier for this press release
     * @param headline the headline of the press release
     * @param body the body content of the press release
     * @param contact the contact information for the press release
     * @param template the template used for rendering this press release
     * @param customFields the custom fields to be used in template substitution
     */
    public PressRelease(String id, String headline, String body, String contact, Template template, java.util.Map<String, String> customFields) {
        this.id = id;
        this.headline = headline;
        this.body = body;
        this.contact = contact;
        this.template = template;
        // Create a defensive copy of the custom fields map to prevent external mutation
        this.customFields = customFields == null ? new java.util.LinkedHashMap<>() : new java.util.LinkedHashMap<>(customFields);
        this.state = ReleaseState.DRAFT;
        this.tags = new java.util.LinkedHashSet<>();
        this.versionHistory = new java.util.ArrayList<>();
    }

    /**
     * Returns the unique identifier of this press release.
     *
     * @return the ID
     */
    public String getId() {
        return id;
    }

    /**
     * Returns the headline of this press release.
     *
     * @return the headline
     */
    public String getHeadline() {
        return headline;
    }

    /**
     * Returns the body content of this press release.
     *
     * @return the body
     */
    public String getBody() {
        return body;
    }

    /**
     * Returns the contact information of this press release.
     *
     * @return the contact
     */
    public String getContact() {
        return contact;
    }

    /**
     * Returns the current state of this press release.
     *
     * @return the state
     */
    public ReleaseState getState() {
        return state;
    }

    /**
     * Validates the content of this press release.
     * The headline, body, and contact must be valid according to the Validator rules.
     *
     * @return true if the content is valid, false otherwise
     */
    public boolean validate() {
        return Validator.validateHeadline(headline) &&
               Validator.validateBody(body) &&
               Validator.validateContact(contact);
    }

    /**
     * Submits this press release for review.
     * The press release must be in DRAFT state and must pass validation.
     *
     * @throws IllegalStateException if the press release is not in DRAFT state or fails validation
     */
    public void submitForReview() {
        if (state != ReleaseState.DRAFT) {
            throw new IllegalStateException("Cannot submit for review: press release is not in DRAFT state");
        }
        if (!validate()) {
            throw new IllegalStateException("Cannot submit for review: press release content is invalid");
        }
        state = ReleaseState.REVIEW;
    }

    /**
     * Publishes this press release.
     * The press release must be in REVIEW state.
     * On first publication, a version is added to the version history.
     *
     * @throws IllegalStateException if the press release is not in REVIEW state
     */
    public void publish() {
        if (state != ReleaseState.REVIEW) {
            throw new IllegalStateException("Cannot publish: press release is not in REVIEW state");
        }
        // Add version history on first publication
        if (versionHistory.isEmpty()) {
            String renderedText = renderPlainText();
            versionHistory.add(new Version(1, renderedText));
        }
        state = ReleaseState.PUBLISHED;
    }

    /**
     * Adds a tag to this press release.
     * Tags are stored in insertion order and duplicates are avoided.
     * Null or blank tags are ignored.
     *
     * @param tag the tag to add
     */
    public void addTag(String tag) {
        if (tag != null && !tag.trim().isEmpty()) {
            tags.add(tag.trim());
        }
    }

    /**
     * Returns a defensive copy of the tags associated with this press release.
     * The tags are returned in insertion order.
     *
     * @return a list of tags
     */
    public java.util.List<String> getTags() {
        return new java.util.ArrayList<>(tags);
    }

    /**
     * Renders this press release as plain text using the assigned template and custom fields.
     * Combines the headline, body, contact, and custom fields into a single map for substitution.
     *
     * @return the rendered plain text
     */
    public String renderPlainText() {
        java.util.Map<String, String> fields = new java.util.LinkedHashMap<>();
        fields.put("headline", headline);
        fields.put("body", body);
        fields.put("contact", contact);
        fields.putAll(customFields);
        return template.applyCustomFields(fields);
    }

    /**
     * Returns a defensive copy of the version history of this press release.
     * The versions are returned in chronological order.
     *
     * @return a list of versions
     */
    public java.util.List<Version> getVersionHistory() {
        return new java.util.ArrayList<>(versionHistory);
    }
}

File: results/RQ1/05_baseline_qualification/units/p11_custom_press_releases/python/replays/0001_seed_2037595148/p11_custom_press_releases_python/p11_custom_press_releases_python/press_releases/press_release.py
"""Mutable press release aggregate with lifecycle management."""

from typing import Dict, List, Optional
from .errors import DomainError, InvalidTransitionError
from .types import State, Contact, VersionEntry
from .template import Template
from .validation import validate_headline, validate_body, validate_contact


class PressRelease:
    """A mutable press release with lifecycle management."""

    def __init__(
        self,
        id: str,
        headline: str,
        body: str,
        contact: Contact,
        template: Template,
        custom_fields: Optional[Dict[str, str]] = None,
    ):
        """Initialize a press release.

        Note: This constructor deliberately does not validate inputs. Invalid
        values are allowed to be stored so that validate() can properly
        determine validity.

        Args:
            id: Unique identifier for the press release.
            headline: The press release headline.
            body: The press release body text.
            contact: The contact person for the press release.
            template: The template to use for rendering.
            custom_fields: Optional custom fields to be used in rendering.
        """
        self._id = id
        self._headline = headline
        self._body = body
        self._contact = contact
        self._template = template
        self._custom_fields = custom_fields if custom_fields is not None else {}
        self._state = State.DRAFT
        self._tags_list: List[str] = []
        self._tags_set: set[str] = set()
        self._versions: List[VersionEntry] = []

    @property
    def id(self) -> str:
        """Get the press release ID."""
        return self._id

    @property
    def headline(self) -> str:
        """Get the press release headline."""
        return self._headline

    @property
    def body(self) -> str:
        """Get the press release body."""
        return self._body

    @property
    def contact(self) -> Contact:
        """Get the press release contact."""
        return self._contact

    @property
    def state(self) -> State:
        """Get the current state of the press release."""
        return self._state

    def validate(self) -> bool:
        """Validate the press release.

        Returns:
            True if all fields are valid, False otherwise.
        """
        return (
            validate_headline(self._headline)
            and validate_body(self._body)
            and validate_contact(self._contact)
        )

    def submit_for_review(self) -> None:
        """Submit the press release for review.

        Raises:
            DomainError: If the press release is invalid.
        """
        if not self.validate():
            raise DomainError("Cannot submit for review: invalid press release")

        if self._state != State.DRAFT:
            raise InvalidTransitionError(
                f"Cannot submit for review: expected DRAFT, got {self._state.value}"
            )

        self._state = State.REVIEW

    def publish(self) -> None:
        """Publish the press release.

        Raises:
            InvalidTransitionError: If the press release is not in REVIEW state.
        """
        if self._state != State.REVIEW:
            raise InvalidTransitionError(
                f"Cannot publish: expected REVIEW, got {self._state.value}"
            )

        self._state = State.PUBLISHED

        # Append version history on first publish
        rendered_text = self.render_plain_text()
        version_entry = VersionEntry(number=1, rendered_text=rendered_text)
        self._versions.append(version_entry)

    def add_tag(self, tag: str) -> None:
        """Add a tag to the press release.

        Args:
            tag: The tag to add.

        Raises:
            DomainError: If the tag is blank.
        """
        # Trim whitespace
        trimmed_tag = tag.strip()
        if not trimmed_tag:
            raise DomainError("Tag cannot be blank")

        # Add only if not already present
        if trimmed_tag not in self._tags_set:
            self._tags_list.append(trimmed_tag)
            self._tags_set.add(trimmed_tag)

    def tags(self) -> List[str]:
        """Get the list of tags.

        Returns:
            A defensive copy of the tags list.
        """
        return self._tags_list.copy()

    def render_plain_text(self) -> str:
        """Render the press release as plain text.

        Combines custom fields with headline, body, and contact information,
        then renders using the template.

        Returns:
            The rendered plain text.
        """
        # Prepare fields for rendering
        fields = self._custom_fields.copy()
        fields["headline"] = self._headline
        fields["body"] = self._body
        fields["contact_name"] = self._contact.name
        fields["contact_email"] = self._contact.email

        return self._template.render(fields)

    def version_history(self) -> List[VersionEntry]:
        """Get the version history.

        Returns:
            A defensive copy of the version history list.
        """
        return self._versions.copy()